File: ropes_bp/scripts/domain/RopeChain.js
import { MAX_CHAIN_LENGTH } from "../util/RopeConstants.js";

export class RopeChain {
  constructor(id, type, dimensionId, anchorPos, anchorFace, initialSegments = 0) {
    this.id = id;
    this.type = type;
    this.dimensionId = dimensionId;
    this.anchorPos = { ...anchorPos };
    this.anchorFace = anchorFace;
    this.isWhipDeployed = false;
    this.deployerName = null;
    this.drops = [
      {
        coilPos: { ...anchorPos },
        remaining: initialSegments,
        segments: [],
      },
    ];
  }

  get totalSegments() {
    let total = 0;
    for (const drop of this.drops) {
      total += drop.remaining + drop.segments.length;
    }
    return total;
  }

  getDrop(index) {
    return this.drops[index] ?? null;
  }

  getDropForPosition(pos) {
    for (let i = 0; i < this.drops.length; i++) {
      const drop = this.drops[i];
      if (drop.coilPos.x === pos.x && drop.coilPos.y === pos.y && drop.coilPos.z === pos.z) {
        return { drop, dropIndex: i, isCoil: true };
      }
      for (let s = 0; s < drop.segments.length; s++) {
        const seg = drop.segments[s];
        if (seg.x === pos.x && seg.y === pos.y && seg.z === pos.z) {
          return { drop, dropIndex: i, isCoil: false, segmentIndex: s };
        }
      }
    }
    return null;
  }

  isAnchor(pos) {
    return this.anchorPos.x === pos.x && this.anchorPos.y === pos.y && this.anchorPos.z === pos.z;
  }

  isCoilPosition(pos) {
    for (const drop of this.drops) {
      if (drop.coilPos.x === pos.x && drop.coilPos.y === pos.y && drop.coilPos.z === pos.z) {
        return true;
      }
    }
    return false;
  }

  addSegments(count, dropIndex = 0) {
    const drop = this.drops[dropIndex];
    if (!drop) return 0;
    const space = MAX_CHAIN_LENGTH - this.totalSegments;
    const added = Math.min(count, space);
    drop.remaining += added;
    return added;
  }

  extendDrop(dropIndex, positions) {
    const drop = this.drops[dropIndex];
    if (!drop) return { extended: 0, remaining: 0 };
    const toExtend = Math.min(positions.length, drop.remaining);
    for (let i = 0; i < toExtend; i++) {
      drop.segments.push({ ...positions[i] });
      drop.remaining--;
    }
    return { extended: toExtend, remaining: drop.remaining };
  }

  createLedgeCoil(dropIndex, coilPos) {
    const parentDrop = this.drops[dropIndex];
    if (!parentDrop || parentDrop.remaining <= 0) return null;
    const newDrop = {
      coilPos: { ...coilPos },
      remaining: parentDrop.remaining,
      segments: [],
    };
    parentDrop.remaining = 0;
    this.drops.splice(dropIndex + 1, 0, newDrop);
    return newDrop;
  }

  retractOneFromBottom() {
    for (let i = this.drops.length - 1; i >= 0; i--) {
      const drop = this.drops[i];
      if (drop.segments.length > 0) {
        const removed = drop.segments.pop();
        drop.remaining++;
        if (drop.segments.length === 0 && i > 0) {
          const parent = this.drops[i - 1];
          parent.remaining += drop.remaining;
          this.drops.splice(i, 1);
          return { removed, ledgeCoilRemoved: drop.coilPos };
        }
        return { removed, ledgeCoilRemoved: null };
      }
    }
    return null;
  }

  fullRecoil() {
    const removedPositions = [];
    for (let i = this.drops.length - 1; i >= 0; i--) {
      const drop = this.drops[i];
      for (const seg of drop.segments) {
        removedPositions.push({ ...seg });
      }
      if (i > 0) {
        removedPositions.push({ ...drop.coilPos });
      }
    }
    const total = this.totalSegments;
    this.drops = [{
      coilPos: { ...this.anchorPos },
      remaining: total,
      segments: [],
    }];
    return removedPositions;
  }

  breakAtSegment(dropIndex, segmentIndex) {
    const brokenPositions = [];
    const drop = this.drops[dropIndex];

    for (let s = segmentIndex; s < drop.segments.length; s++) {
      brokenPositions.push({ ...drop.segments[s] });
    }
    drop.segments.length = segmentIndex;

    for (let i = this.drops.length - 1; i > dropIndex; i--) {
      const d = this.drops[i];
      for (const seg of d.segments) {
        brokenPositions.push({ ...seg });
      }
      brokenPositions.push({ ...d.coilPos });
    }
    this.drops.length = dropIndex + 1;

    return brokenPositions;
  }

  breakAll() {
    const positions = [];
    for (const drop of this.drops) {
      for (const seg of drop.segments) {
        positions.push({ ...seg });
      }
      if (drop.coilPos.x !== this.anchorPos.x ||
          drop.coilPos.y !== this.anchorPos.y ||
          drop.coilPos.z !== this.anchorPos.z) {
        positions.push({ ...drop.coilPos });
      }
    }
    positions.push({ ...this.anchorPos });
    return positions;
  }

  getAllPositions() {
    const positions = [{ ...this.anchorPos }];
    for (const drop of this.drops) {
      if (drop.coilPos.x !== this.anchorPos.x ||
          drop.coilPos.y !== this.anchorPos.y ||
          drop.coilPos.z !== this.anchorPos.z) {
        positions.push({ ...drop.coilPos });
      }
      for (const seg of drop.segments) {
        positions.push({ ...seg });
      }
    }
    return positions;
  }

  toJSON() {
    return {
      id: this.id,
      type: this.type,
      dimensionId: this.dimensionId,
      anchorPos: this.anchorPos,
      anchorFace: this.anchorFace,
      isWhipDeployed: this.isWhipDeployed,
      deployerName: this.deployerName,
      drops: this.drops.map(d => ({
        coilPos: d.coilPos,
        remaining: d.remaining,
        segments: d.segments.map(s => ({ x: s.x, y: s.y, z: s.z })),
      })),
    };
  }

  static fromJSON(json) {
    const chain = new RopeChain(
      json.id,
      json.type,
      json.dimensionId,
      json.anchorPos,
      json.anchorFace,
      0,
    );
    chain.isWhipDeployed = json.isWhipDeployed ?? false;
    chain.deployerName = json.deployerName ?? null;
    chain.drops = json.drops.map(d => ({
      coilPos: { ...d.coilPos },
      remaining: d.remaining,
      segments: d.segments.map(s => ({ x: s.x, y: s.y, z: s.z })),
    }));
    return chain;
  }
}

File: java-ropes/src/main/java/com/ropes/domain/RopeChain.java
package com.ropes.domain;

import com.ropes.util.RopeConstants;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Pure game logic for a single rope/rope-ladder/whip chain. No Minecraft
 * imports — this is a faithful port of scripts/domain/RopeChain.js and is
 * exercised by RopeChainTest with plain JUnit.
 */
public class RopeChain {
    private final String id;
    private final String type;
    private final String dimensionId;
    private final RopePos anchorPos;
    private final String anchorFace;
    private boolean whipDeployed = false;
    private String deployerName = null;
    private List<Drop> drops = new ArrayList<>();

    public RopeChain(String id, String type, String dimensionId, RopePos anchorPos,
                     String anchorFace, int initialSegments) {
        this.id = id;
        this.type = type;
        this.dimensionId = dimensionId;
        this.anchorPos = anchorPos;
        this.anchorFace = anchorFace;
        this.drops.add(new Drop(anchorPos, initialSegments));
    }

    public String getId() { return id; }
    public String getType() { return type; }
    public String getDimensionId() { return dimensionId; }
    public RopePos getAnchorPos() { return anchorPos; }
    public String getAnchorFace() { return anchorFace; }
    public boolean isWhipDeployed() { return whipDeployed; }
    public void setWhipDeployed(boolean whipDeployed) { this.whipDeployed = whipDeployed; }
    public String getDeployerName() { return deployerName; }
    public void setDeployerName(String deployerName) { this.deployerName = deployerName; }
    public List<Drop> getDrops() { return drops; }

    public int getTotalSegments() {
        int total = 0;
        for (Drop drop : drops) {
            total += drop.getRemaining() + drop.getSegments().size();
        }
        return total;
    }

    public Drop getDrop(int index) {
        if (index < 0 || index >= drops.size()) return null;
        return drops.get(index);
    }

    public DropLocation getDropForPosition(RopePos pos) {
        for (int i = 0; i < drops.size(); i++) {
            Drop drop = drops.get(i);
            if (drop.getCoilPos().equals(pos)) {
                return new DropLocation(drop, i, true, -1);
            }
            List<RopePos> segments = drop.getSegments();
            for (int s = 0; s < segments.size(); s++) {
                if (segments.get(s).equals(pos)) {
                    return new DropLocation(drop, i, false, s);
                }
            }
        }
        return null;
    }

    public boolean isAnchor(RopePos pos) {
        return anchorPos.equals(pos);
    }

    public boolean isCoilPosition(RopePos pos) {
        for (Drop drop : drops) {
            if (drop.getCoilPos().equals(pos)) return true;
        }
        return false;
    }

    public int addSegments(int count) {
        return addSegments(count, 0);
    }

    public int addSegments(int count, int dropIndex) {
        Drop drop = getDrop(dropIndex);
        if (drop == null) return 0;
        int space = RopeConstants.MAX_CHAIN_LENGTH - getTotalSegments();
        int added = Math.min(count, space);
        drop.addRemaining(added);
        return added;
    }

    public ExtendResult extendDrop(int dropIndex, List<RopePos> positions) {
        Drop drop = getDrop(dropIndex);
        if (drop == null) return new ExtendResult(0, 0);
        int toExtend = Math.min(positions.size(), drop.getRemaining());
        for (int i = 0; i < toExtend; i++) {
            drop.getSegments().add(positions.get(i));
            drop.setRemaining(drop.getRemaining() - 1);
        }
        return new ExtendResult(toExtend, drop.getRemaining());
    }

    public Drop createLedgeCoil(int dropIndex, RopePos coilPos) {
        Drop parentDrop = getDrop(dropIndex);
        if (parentDrop == null || parentDrop.getRemaining() <= 0) return null;
        Drop newDrop = new Drop(coilPos, parentDrop.getRemaining());
        parentDrop.setRemaining(0);
        drops.add(dropIndex + 1, newDrop);
        return newDrop;
    }

    public RetractResult retractOneFromBottom() {
        for (int i = drops.size() - 1; i >= 0; i--) {
            Drop drop = drops.get(i);
            List<RopePos> segments = drop.getSegments();
            if (!segments.isEmpty()) {
                RopePos removed = segments.remove(segments.size() - 1);
                drop.addRemaining(1);
                if (segments.isEmpty() && i > 0) {
                    Drop parent = drops.get(i - 1);
                    parent.addRemaining(drop.getRemaining());
                    RopePos ledgeCoil = drop.getCoilPos();
                    drops.remove(i);
                    return new RetractResult(removed, ledgeCoil);
                }
                return new RetractResult(removed, null);
            }
        }
        return null;
    }

    public List<RopePos> fullRecoil() {
        List<RopePos> removedPositions = new ArrayList<>();
        for (int i = drops.size() - 1; i >= 0; i--) {
            Drop drop = drops.get(i);
            removedPositions.addAll(drop.getSegments());
            if (i > 0) {
                removedPositions.add(drop.getCoilPos());
            }
        }
        int total = getTotalSegments();
        drops = new ArrayList<>();
        drops.add(new Drop(anchorPos, total));
        return removedPositions;
    }

    public List<RopePos> breakAtSegment(int dropIndex, int segmentIndex) {
        List<RopePos> brokenPositions = new ArrayList<>();
        Drop drop = drops.get(dropIndex);
        List<RopePos> segments = drop.getSegments();

        for (int s = segmentIndex; s < segments.size(); s++) {
            brokenPositions.add(segments.get(s));
        }
        while (segments.size() > segmentIndex) {
            segments.remove(segments.size() - 1);
        }

        for (int i = drops.size() - 1; i > dropIndex; i--) {
            Drop d = drops.get(i);
            brokenPositions.addAll(d.getSegments());
            brokenPositions.add(d.getCoilPos());
        }
        while (drops.size() > dropIndex + 1) {
            drops.remove(drops.size() - 1);
        }

        return brokenPositions;
    }

    public List<RopePos> breakAll() {
        List<RopePos> positions = new ArrayList<>();
        for (Drop drop : drops) {
            positions.addAll(drop.getSegments());
            if (!drop.getCoilPos().equals(anchorPos)) {
                positions.add(drop.getCoilPos());
            }
        }
        positions.add(anchorPos);
        return positions;
    }

    public List<RopePos> getAllPositions() {
        List<RopePos> positions = new ArrayList<>();
        positions.add(anchorPos);
        for (Drop drop : drops) {
            if (!drop.getCoilPos().equals(anchorPos)) {
                positions.add(drop.getCoilPos());
            }
            positions.addAll(drop.getSegments());
        }
        return positions;
    }

    public Map<String, Object> toMap() {
        Map<String, Object> map = new LinkedHashMap<>();
        map.put("id", id);
        map.put("type", type);
        map.put("dimensionId", dimensionId);
        map.put("anchorPos", anchorPos.toMap());
        map.put("anchorFace", anchorFace);
        map.put("isWhipDeployed", whipDeployed);
        map.put("deployerName", deployerName);
        List<Object> dropList = new ArrayList<>();
        for (Drop drop : drops) {
            Map<String, Object> dropMap = new LinkedHashMap<>();
            dropMap.put("coilPos", drop.getCoilPos().toMap());
            dropMap.put("remaining", drop.getRemaining());
            List<Object> segs = new ArrayList<>();
            for (RopePos seg : drop.getSegments()) {
                segs.add(seg.toMap());
            }
            dropMap.put("segments", segs);
            dropList.add(dropMap);
        }
        map.put("drops", dropList);
        return map;
    }

    @SuppressWarnings("unchecked")
    public static RopeChain fromMap(Map<String, Object> json) {
        RopeChain chain = new RopeChain(
            (String) json.get("id"),
            (String) json.get("type"),
            (String) json.get("dimensionId"),
            RopePos.fromMap((Map<String, Object>) json.get("anchorPos")),
            (String) json.get("anchorFace"),
            0
        );
        Object whip = json.get("isWhipDeployed");
        chain.whipDeployed = whip instanceof Boolean && (Boolean) whip;
        chain.deployerName = (String) json.get("deployerName");

        List<Drop> newDrops = new ArrayList<>();
        List<Map<String, Object>> dropMaps = (List<Map<String, Object>>) json.get("drops");
        for (Map<String, Object> dropMap : dropMaps) {
            RopePos coil = RopePos.fromMap((Map<String, Object>) dropMap.get("coilPos"));
            int remaining = ((Number) dropMap.get("remaining")).intValue();
            Drop drop = new Drop(coil, remaining);
            List<Map<String, Object>> segMaps = (List<Map<String, Object>>) dropMap.get("segments");
            for (Map<String, Object> segMap : segMaps) {
                drop.getSegments().add(RopePos.fromMap(segMap));
            }
            newDrops.add(drop);
        }
        chain.drops = newDrops;
        return chain;
    }

    /** Result of {@link #getDropForPosition}; segmentIndex is -1 for a coil hit. */
    public record DropLocation(Drop drop, int dropIndex, boolean isCoil, int segmentIndex) {}

    /** Result of {@link #extendDrop}. */
    public record ExtendResult(int extended, int remaining) {}

    /** Result of {@link #retractOneFromBottom}; ledgeCoilRemoved is null when no drop merged. */
    public record RetractResult(RopePos removed, RopePos ledgeCoilRemoved) {}
}
